src/arch/aarch64/Emit.zig

```diff
@@ -654,15 +654,31 @@ fn mirLoadMemoryPie(emit: *Emit, inst: Mir.Inst.Index) !void {
     const data = emit.mir.extraData(Mir.LoadMemoryPie, payload).data;
     const reg = @intToEnum(Register, data.register);
 
-    // PC-relative displacement to the entry in the GOT table.
+    // PC-relative displacement to the entry in memory.
     // adrp
     const offset = @intCast(u32, emit.code.items.len);
     try emit.writeInstruction(Instruction.adrp(reg, 0));
 
     switch (tag) {
-        .load_memory_got,
-        .load_memory_direct,
-        => {
+        .load_memory_got => {
+            // ldr reg, reg, offset
+            try emit.writeInstruction(Instruction.ldr(
+                reg,
+                reg,
+                Instruction.LoadStoreOffset.imm(0),
+            ));
+        },
+        .load_memory_direct => {
+            // We cannot load the offset directly as it may not be aligned properly.
```

Comment:
I think another long-term solution for this is to emit `ldrb`, `ldrh`, `ldrsb`, and `ldrsh` when appropriate. While `ldr` requires the address to be aligned to 8 bytes, `ldrh`/`ldrsh` require an alignment of 2 and `ldrb`/`ldrsb` require an alignment of 1 (no alignment).

Of course, these instructions are meant for loading integers which have natural alignment. For types with different alignments, we can fall back to the `adrp, add, ldr` or have the linker guarantee larger alignment so the `add` is not necessary anymore.

But for now, this is a perfectly fine solution.

Reply:
Agreed on different `ldr` variants. As for guaranteeing larger alignment by the linker, that's a no-go: the linker should only ever put the decl/const at the required alignment. However, it can optimise the `add` out and replace with an `ldr` if the target address satisfies the alignment requirements of the `ldr` instruction. It can then also `nop` the 3rd `ldr` in that case; however note it won't be able to remove it as this would potentially destroy subsequent relocations as the offsets would change potentially resulting in misaligned accesses etc.

Reply:
> Agreed on different `ldr` variants. As for guaranteeing larger alignment by the linker, that's a no-go: the linker should only ever put the decl/const at the required alignment.

That makes sense, yes.

> However, it can optimise the `add` out and replace with an `ldr` if the target address satisfies the alignment requirements of the `ldr` instruction. It can then also `nop` the 3rd `ldr` in that case; however note it won't be able to remove it as this would potentially destroy subsequent relocations as the offsets would change potentially resulting in misaligned accesses etc.

Agreed. After Emit, we shouldn't be allowed to delete instructions.